File: backend/src/k8s-monitor.js
const k8s = require('@kubernetes/client-node');
const fs = require('fs');
const path = require('path');
const https = require('https');

class K8sMonitor {
  constructor() {
    try {
      const kc = new k8s.KubeConfig();
      
      const kubeconfigPath = path.join(process.env.HOME || '/root', '.kube', 'config');
      const k3dConfigPath = path.join(process.env.HOME || '/root', '.kube', 'k3d-config');
      
      if (fs.existsSync(kubeconfigPath)) {
        kc.loadFromFile(kubeconfigPath);
      } else if (fs.existsSync(k3dConfigPath)) {
        kc.loadFromFile(k3dConfigPath);
      } else {
        kc.loadFromDefault();
      }
      
      // Create clients with proper agent
      const agent = new https.Agent({
        rejectUnauthorized: false,
        checkServerIdentity: () => undefined
      });
      
      this.api = kc.makeApiClient(k8s.CoreV1Api);
      this.appsApi = kc.makeApiClient(k8s.AppsV1Api);
      
      // Set agent on the request object
      if (this.api && this.api.setDefaultAuthentication) {
        const opts = kc.getCurrentCluster();
        if (opts) {
          opts.skipTLSVerify = true;
        }
      }
      
      this.kc = kc;
      this.agent = agent;
      this.ready = true;
    } catch (error) {
      console.warn('⚠️ K8s init:', error.message);
      this.api = null;
      this.ready = false;
    }
  }

  async getClusterInfo() {
    try {
      if (!this.api || !this.ready) return { error: 'K8s unavailable' };
      const nodes = await this.api.listNode();
      return {
        nodes: nodes.body.items.length,
        nodeList: nodes.body.items.map(n => ({
          name: n.metadata.name,
          status: n.status.conditions.find(c => c.type === 'Ready')?.status
        }))
      };
    } catch (error) {
      return { error: error.message };
    }
  }

  async getAllPods() {
    try {
      if (!this.api || !this.ready) return [];
      const res = await this.api.listPodForAllNamespaces();
      console.log('K8s pods fetched:', res.body.items.length);
      return res.body.items.map(p => ({
        name: p.metadata.name,
        namespace: p.metadata.namespace,
        status: p.status.phase,
        containers: p.spec.containers.length,
        image: p.spec.containers[0]?.image
      }));
    } catch (error) {
      console.error('K8s pods error:', error.message);
      return [];
    }
  }
}

module.exports = K8sMonitor;

File: backend/src/k8s_monitor.py
from kubernetes import client, config
from kubernetes.client.rest import ApiException
import os
from pathlib import Path

class K8sMonitor:
    def __init__(self):
        try:
            kubeconfig_path = os.path.join(os.path.expanduser("~"), ".kube", "config")
            k3d_config_path = os.path.join(os.path.expanduser("~"), ".kube", "k3d-config")
            
            if Path(kubeconfig_path).exists():
                config.load_kube_config(kubeconfig_path)
            elif Path(k3d_config_path).exists():
                config.load_kube_config(k3d_config_path)
            else:
                config.load_incluster_config()
            
            self.v1 = client.CoreV1Api()
            self.apps_v1 = client.AppsV1Api()
            self.ready = True
            print("✅ K8s monitoring initialized")
        except Exception as e:
            print(f"⚠️ K8s init: {str(e)}")
            self.v1 = None
            self.ready = False

    def get_cluster_info(self):
        try:
            if not self.ready or not self.v1:
                return {"error": "K8s unavailable"}
            
            nodes = self.v1.list_node()
            return {
                "nodes": len(nodes.items),
                "nodeList": [
                    {
                        "name": node.metadata.name,
                        "status": next(
                            (c.status for c in node.status.conditions if c.type == "Ready"),
                            "Unknown"
                        ),
                        "kubeletVersion": node.status.node_info.kubelet_version
                    }
                    for node in nodes.items
                ]
            }
        except Exception as e:
            return {"error": str(e)}

    def get_all_pods(self):
        try:
            if not self.ready or not self.v1:
                return []
            
            pods = self.v1.list_pod_for_all_namespaces()
            print(f"K8s pods fetched: {len(pods.items)}")
            return [
                {
                    "name": pod.metadata.name,
                    "namespace": pod.metadata.namespace,
                    "status": pod.status.phase,
                    "containers": len(pod.spec.containers),
                    "image": pod.spec.containers[0].image if pod.spec.containers else "N/A"
                }
                for pod in pods.items
            ]
        except Exception as e:
            print(f"K8s pods error: {str(e)}")
            return []
